mijia-homie/src/config.rs

```diff
@@ -0,0 +1,128 @@
+use eyre::Report;
+use mijia::bluetooth::{MacAddress, ParseMacAddressError};
+use rumqttc::MqttOptions;
+use rustls::ClientConfig;
+use serde_derive::Deserialize;
+use stable_eyre::eyre::WrapErr;
+use std::collections::HashMap;
+use std::fs::read_to_string;
+use std::sync::Arc;
+
+const DEFAULT_MQTT_PREFIX: &str = "homie";
+const DEFAULT_DEVICE_ID: &str = "mijia-bridge";
+const DEFAULT_DEVICE_NAME: &str = "Mijia bridge";
+const DEFAULT_HOST: &str = "test.mosquitto.org";
+const DEFAULT_PORT: u16 = 1883;
+const DEFAULT_SENSOR_NAMES_FILENAME: &str = "sensor_names.toml";
+const CONFIG_FILENAME: &str = "mijia_homie.toml";
+
+#[derive(Clone, Debug, Default, Deserialize)]
+#[serde(default, deny_unknown_fields)]
+pub struct Config {
+    pub mqtt: MqttConfig,
+    pub homie: HomieConfig,
+}
+
+impl Config {
+    pub fn from_file() -> Result<Config, Report> {
+        Config::read(CONFIG_FILENAME)
+    }
+
+    fn read(filename: &str) -> Result<Config, Report> {
+        let config_file =
+            read_to_string(filename).wrap_err_with(|| format!("Reading {}", filename))?;
+        Ok(toml::from_str(&config_file)?)
+    }
+}
+
+#[derive(Clone, Debug, Deserialize)]
+#[serde(default, deny_unknown_fields)]
+pub struct MqttConfig {
+    pub host: String,
+    pub port: u16,
+    pub use_tls: bool,
+    pub username: Option<String>,
+    pub password: Option<String>,
+    pub client_name: Option<String>,
+}
+
+impl Default for MqttConfig {
+    fn default() -> MqttConfig {
+        MqttConfig {
+            host: DEFAULT_HOST.to_owned(),
+            port: DEFAULT_PORT,
+            use_tls: false,
+            username: None,
+            password: None,
+            client_name: None,
+        }
+    }
+}
+
+#[derive(Clone, Debug, Deserialize)]
+#[serde(default, deny_unknown_fields)]
+pub struct HomieConfig {
+    pub device_id: String,
+    pub device_name: String,
+    pub prefix: String,
+    pub sensor_names_filename: String,
```

Comment:
add this to mijia-homie/mijia_homie.example.toml ?

Reply:
Done.